### ISSUE TITLE
More explicit selection of Explicit tests

### ISSUE THREAD

**Issue by philipcarlsen** [nunit/nunit-console#134]
I find the semantics of running explicit tests in the nunit-console runner confusing.

From reading the documentation, I gather that filtering on categories can trigger the execution of explicit tests, depending on the given composition of filters, with the idea being that positively selecting a category (like in the filter expression "cat = SomeCategory") implies selecting explicit-marked tests of that category as well.

Our main use cases for explicit-marking is to express tests that either take a lot of resources to execute, or tests that have side effects that need to be controlled for prior to running the tests.

Another (ab)use of explicit tests is the convenient updating of expectations stored in the file system, eg. for tests that ensure that some generated value or file remains constant in the face of unrelated changes.

As a consequence we're interested in filtering tests based on categories in the console runner, while delegating the running of explicit-marked tests to other, more interactive test runners, such as that of ReSharper.

I suppose that a workaround would be for us to place category-attributes on the individual test methods. That however greatly reduces the usefulness of the ability to put category-attributes on test fixture classes.

Also, as the expressiveness of the Test Selection Language increases it becomes less obvious what constitutes a positive selection of some subset of tests. Does "!(cat != SomeCategory)" constitute a positive selection of "SomeCategory"-tests? Should it?

In my opinion the ideal solution would be to have the console runner consider explicit tests as a completely distinct set of tests that would only be taken into consideration for running and filtering when the runner is given a separate "--explicit" switch. That way the complex interaction between filters and explicitness disappears, confusion is reduced, a more expressive selection mechanism is gained, and profit and happiness ensues :-)

What are your thoughts on this?

**Comment by CharliePoole** [comment]
Explicit has always been a bit confusing. Originally designed as s property of a category, it quickly changed to an independent attribute with the meaning "only run on positive selection."

It's most useful when that selection is by name or by direct selection in the gui. As you say, the evolution of TSL has made things more complicated.

We can't really change the underlying meaning of Explicit without breaking compatibility and we can't make it work differently from runner to runner since filters are applied by the framework.

It is possible to create a new filter that includes all explicit tests, essentially treating Explicit as a pseudo-category. I'm not sure this would be generally useful, since most people don't seem to use Explicit in the way you describe.

Obviously, we can and should fix any bugs in how the direct selection rule is applied.

My general advice has always been to reserve explicit for tests you never run in CI but which are run one at a time by dev.

**Comment by daflame76** [comment]
@CharliePoole I know this conversation is getting dated, but I was wondering how you advise to separate CI tests which can be long-running and could involve resources that may not be available on a dev machine?  This obviously isn't a unit tests per se, but an integration or acceptance test.

Currently we are trying to implement some fashion of integration tests which involve components which can be exercised by unit test, but we want to ensure changing them doesn't break integration.

**Comment by daflame76** [comment]
I am bringing this up because I noticed that use of the where clause `--where "cat!=Integration || cat==Integration"` actually includes all explicit tests, not just the ones in the Integration category.

**Comment by CharliePoole** [comment]
@daflame76 In my own work, I use Explicit for things that should never run in CI - it doesn't matter whether they are unit-, integration- or acceptance-tests. That's an easy choice for me, because that's what I created Explicit for! A typical use is a test that always fails and which I run manually in order to check that the message displays as I want. Of course, I will likely have a unit test that serves the same purpose, but I like to actually see the message from time to time.

For acceptance tests, I always have a separate assembly, so the issue doesn't arise. In my definition, acceptance tests fail at the start of a project, so it's not very useful to have them in the "must pass" part of your build script. Generally, I also put large-scale integration tests in a separate assembly, but smaller tests where a few components are cooperating sometimes go into the unit-testing suite. In that case, I like to use a category for them.

We do have some outstanding issues regarding running Explicit tests in the presence of a where-clause with negative entries. Your example is particularly interesting!

**Comment by daflame76** [comment]
I know that where clause is a bit of a kludge, but technically speaking it should select all non-integration tests that are not explicit and any integration tests regardless whether they are explicit or not.

It just seems like there should be a separate query parameter to select explicit tests to avoid the confusion that everyone is suffering.

**Comment by daflame76** [comment]
@CharliePoole if I run each part of the where clause separately that is exactly the behaviour I get.

**Comment by CharliePoole** [comment]
@daflame76 
You are right, in designing Explicit, the intention was:
 * The lhs expression should not select any explicit tests because it's negative (an exclude).
 * The rhs expression should only select explicit tests that have the category.
 * The or should not change anything WRT explicit tests.
From what you are saying, the or expression __is__ changing things.

I believe what is happening is that old (V2) NUnit didn't have the syntax to easily form complex expressions: You only had the `--include` and `--exclude` parameters available. Now we are seeing much more complex expressions and holes are showing up in the implementation. Ironically, we need more tests!

I think some special form to include/exclude explicit tests at the top level of the where clause is desirable but I also think we should fix the bug that is demonstrated by your example.

**Comment by daflame76** [comment]
@CharliePoole if there was an include/exclude explicit at the top level of the expression I think that would simplify the behaviour.  It seems to me there is a run exclusive flag being turned on when the `Integration` category positive match is added in, but it is affecting both sides of the expression.  If the explicit selector was a component of the expression then it would function as part of the match expression rather than a flag that is either on or off.

**Comment by CharliePoole** [comment]
I agree with your idea myself, but I also think the old behavior without the new switch has to be fixed. It has to do with the way the binary operator is combining the explicit selection flag from the lhs and rhs. I'll create a separate issue for this in the nunit repo.

@rprouse This has been an idea for discussion for quite a long time. I suggest we put it on the backlog and flag it as design so somebody can come up with a user-facing syntax. [I'm trying to school myself to stop accepting changes myself!]

**Comment by daflame76** [comment]
@CharliePoole and @rprouse my proposal to maintain backward compatibility would be to allow explicit to be controlled by the where expression either:
1. When it is included in the where expression
1. Or when a switch like `--enable-explicit-filtering` is turned on

That way it can continue with the legacy behaviour unless requested.

**Comment by CharliePoole** [comment]
What I meant in terms of design is exactly __how__ the syntax of the where clause should be changed to incorporate explicit. Personally, I'm less enthusiastic at a switch, which would globally change how NUnit works.

**Comment by CharliePoole** [comment]
@daflame76 The current behavior is not precisely what I would call "legacy behavior" - I think the actual term is "bug". 😄

**Comment by daflame76** [comment]
Well, the current misbehaviour with regards to the example above is definitely a "bug", yes. :smile: 
I was not proposing to keep that misbehaviour.

I agree, adding switches to change behaviour does make things messy.  It was a suggestion if you needed a way to switch between the new query system with explicit controlled specifically by `--where` and the older one where it is controlled by inclusion of categories.  So category selection doesn't imply explicit test selection.

**Comment by CharliePoole** [comment]
@daflame76 I think I confused the issue. The old include/exclude switches have not been supported for a while. I only mentioned them because I think that the new syntax offers users more opportunity to trigger any hidden bugs like this, which is why we are seeing them now. They were probably always there.

**Comment by daflame76** [comment]
@CharliePoole no, I am not confused.  I do clearly understand that the `/include` and `/exclude` switches are gone and backend converted to the new `--where` syntax.  I was referring to the new `--where` syntax with a new root selector for "explicit".

If "explicit" becomes part of the `--where` selector syntax, including categories will still turn the explicit mode on if you maintain compatibility with the current interface.  This would create lots of problems.

The switch I proposed would turn off the implicit enable the "explict" tests so it can only be included if specifically requested by the `--where` query.

**Comment by daflame76** [comment]
Man that last sentence hurt my brain to unfold.  :smile:

**Comment by CharliePoole** [comment]
Yeah... I'll have to mull that over.

**Comment by daflame76** [comment]
Suggestion:
If the current behaviour with categories enabling "explicit" tests is planned to be deprecated, the `--enable-explicit-filtering` switch could become the recommended mode.  If the switch isn't used and selecting a category implicitly enables "explicit" tests, a warning can be provided to let the users know that "'explicit' tests will be enabled" with a disclaimer that this behaviour will be changed in the next major release.

A major version release later the default behaviour could be reversed and offer `--disable-explicit-filtering` instead as a way to restore the deprecated behaviour for compatibility reasons.  Still necessary for some very old projects I would think.

**Comment by daflame76** [comment]
@CharliePoole both the concept and naming of the switch is up to you and I am just offering a possible path.

**Comment by CharliePoole** [comment]
@daflame76 Up to now, we haven't had any discussion of actually changing the behavior whereby category selection includes explicit tests. At least, if you were suggesting it, I didn't understand.

So... the definition of Explicit is the test is only run when directly selected.

In the framework, we can't "look at" the command-line or at what gui menu was selected. We can only see the filters. Up to now, the framework has considered things to be directly selected if any non-negative filter selected them. I guess you are suggesting we exclude category filters from that?

That's certainly possible to implement, since each filter has it's own definition of "directly selected" built in. The question is do we want to do that? I don't think I do, since it's behavior that has existed since NUnit 2.3. Others may feel differently. @nunit/framework-team please respond.

Whether we allow you to re-enable this behavior after removing it via a switch or some other means is another matter. First we have to decide about the behavior change you are suggesting.

Note that we can't remove this behavior entirely. A runner, e.g. a gui, has to have a way to pass a selected test to the framework when the user has directly chosen to execute that test. So, at a minimum the id filter and probably the test name filter have to continue with this behavior. Category, Property, Class and Method filters are up for debate.

**Comment by jnm2** [comment]
Leaving aside backwards compatibility for the moment, I find it unintuitive and a bit frustrating for any non-negative filter to cause an explicit test to run unless the entire filter applies exclusively to explicit tests or applies exclusively to a single test.

Things I expect to run explicit tests:
```
Test == Fully.Qualified.Test.Name(3)
Explicit == true
```

Things I expect to run explicit tests, only if the class itself is marked explicit (or perhaps if ever test in the class is marked explicit?):
```
Class == SomeClass
```

Things I expect to not run explicit tests, even if they just happen to match only explicit tests:
```
Test =~ SomeRegex
Method == SomeMethod
Class =~ SomeRegex
Category == SomeCategory
Everything else
```

I'm providing that as a data point, not necessarily as what I think it should be. I dislike that subjective interpretation of the question. How can you tell how "direct" a selector is, with respect to explicitness? I wish there was another way to solve this that didn't involve special-casing certain kinds of selectors. We're basically trying to reverse-engineer implicit intent from the filter. Is there a way we could make it explicit?

**Comment by jnm2** [comment]
Still leaving aside backwards compatibility for the moment, I wonder how good it would be if `--where` only added unexplicit tests and `--where-explicit` only added explicit tests?

**Comment by CharliePoole** [comment]
I think you are on the right track. For this issue - which is a console issue - the question is what should the user get back given certain input.The framework should not have to reverse engineer the intention, it should be told what to do. How we do that - how we pass the request to the framework - is an implementation detail.

**Comment by daflame76** [comment]
I agree with you @jnm2 as well...  That interface also makes sense to me, however there will still be issues with ratifying situations where both `--where` clauses are being used simultaneously.

**Comment by rprouse** [comment]
Funny, I actually feel the opposite to @jnm2 in that I prefer the current implementation of explicit, but that likely stems from the fact that I treat explicit tests as a rare case to only be used where needed. I also don't use many advanced where clauses to slice and dice my tests. I especially disagree with things like not running explicit tests for `Category == SomeCategory` because that is exactly how I would run explicit tests, I would put all of them in specific categories that allowed me to run them easily.

We need to be careful because many users have come to expect the current behavior and their numbers are hard to gauge compared to the people who expect otherwise because they are not the ones commenting on issues like this.

**Comment by CharliePoole** [comment]
In fact, Explicit started out as a property of categories, before it became a separate attribute. I feel like any change we make to the basic behavior has to be in the 4.0 framework. For this major release, we could conceivably add something like `test is explicit` to the selection language. IIRC `is` has been mentioned before as a possible special operator.

**Comment by jnm2** [comment]
I agree that we can't change this in 3.x. But we could consider adding something, like `--where-explicit`.

**Comment by daflame76** [comment]
Agreed.  I would expect this as a MAJOR change, if it was implemented at all.

**Comment by daflame76** [comment]
@CharliePoole sounds like a good idea to me.

**Comment by CharliePoole** [comment]
I'll make this an implementation proposal, for current NUnit, without regard to any future major version changes:

1. Explicit selection will continue to work as designed, with any bugs fixed as needed.

2. We'll add the operator `is` to the existing list of operators specified in https://github.com/nunit/docs/wiki/Test-Selection-Language 

3. The following expressions will be accepted, fixing #179

```
cat is present
cat is absent
```

4. The following expressions will be accepted, fixing the current issue

```
test is explicit
```

5. Expressions using `is` will be usable with the normal binary operators (`&&`, `||` and their aliases) as well as the `!` or `not` operator. Thus, anyone wanting the (non-standard) behavior whereby use of a category fails to include explicit tests may use an expression like...

```
--where cat == SomeValue && !(test is explicit)
```

6. Item 5 could be simplified if we also added a `not` operator, i.e.

```
--where cat == SomeValue && test not explicit
```

We could look into this, but I would not want to see it implemented if it complicated the parsing too much. We already have a not prefix operator so a not binary operator introduces context sensitivity to the language. I hesitated to add this as a suggestion, because I fear the tendency of programmers to take such a statement as a challenge. 😄 However, if anyone wants to take it up, please consider the challenge as not being about implementing it - any beginner at parsing can do that - but in implementing it simply. 😈 

**Note:** This is one of those annoying things that requires implementation in both the engine and the framework. We need some minimal design, which would include some version of what I've already written plus the definition of the xml representation of the filters that are generated as a result. I assume that a mod of category filter plus a new filter for explicit will be needed, but I haven't really looked at it beyond the language side of things, which I think we have to decide first.

@nunit/engine-team @nunit/framework-team Comments on the proposal?

**Comment by CharliePoole** [comment]
@rprouse I think this is ready to be a backlog item as an enhancement with the design label. I haven't done anything with it because we are changing how we operate and you haven't yet told the @nunit/framework-team and @nunit/engine-team how you want such decisions handled.

[This might seem kind of silly of me - treating you as if you are a new lead. OTOH, I think it's an opportunity for you to do a reset and tell people how you want things done in these two projects.]

**Comment by rprouse** [comment]
@CharliePoole I like your proposal, it seems like a good balance of maintaining compatibility while acknowledging people's concerns with explicit test selection.

As for resetting, I agree that this is a good chance. I was partially waiting for the full transition and I am also swamped at work right now and just back from vacation, so I haven't had time to think much about it yet.

**Comment by felixpelletier** [comment]
Hey there! 
Was anything ever implemented for this? 
Our team wants to filter out Explicit tests as well. We filter our tests on a category and, through some unfortunate inheritance hierarchy, these Explicit test fit in the filter.
We use "Explicit" for tests that are run only a few times on dev machines exclusively, pretty much as @CharliePoole described. They are bound to fail on CI.

Something like `test not explicit` would be really helpful to us.

**Comment by CharliePoole** [comment]
Rereading this after so many years, I'm surprised it's under the nunit-console project. It seems like most, if not all of the work needs to be done in the framework project. Of course, back when it originated, the two projects had the same lead so it was less important.

@ChrisMaddock would you agree that this needs to be transferred? I'm thinking it may get more attention in the framework project, where the code is.
